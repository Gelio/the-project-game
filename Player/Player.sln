
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 15
VisualStudioVersion = 15.0.27004.2005
MinimumVisualStudioVersion = 10.0.40219.1
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Player", "Player\Player.csproj", "{FAF0048B-015B-4288-93A8-8FE5BA3211E0}"
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Player.Tests", "Player.Tests\Player.Tests.csproj", "{A6990F7B-B90E-46EE-AE6A-5B8053A1D929}"
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Any CPU = Debug|Any CPU
		Release|Any CPU = Release|Any CPU
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{FAF0048B-015B-4288-93A8-8FE5BA3211E0}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{FAF0048B-015B-4288-93A8-8FE5BA3211E0}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{FAF0048B-015B-4288-93A8-8FE5BA3211E0}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{FAF0048B-015B-4288-93A8-8FE5BA3211E0}.Release|Any CPU.Build.0 = Release|Any CPU
		{A6990F7B-B90E-46EE-AE6A-5B8053A1D929}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{A6990F7B-B90E-46EE-AE6A-5B8053A1D929}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{A6990F7B-B90E-46EE-AE6A-5B8053A1D929}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{A6990F7B-B90E-46EE-AE6A-5B8053A1D929}.Release|Any CPU.Build.0 = Release|Any CPU
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
	GlobalSection(ExtensibilityGlobals) = postSolution
		SolutionGuid = {0AA96D8E-CE50-4D1F-ADA4-1E967E9A7F59}
	EndGlobalSection
EndGlobal
