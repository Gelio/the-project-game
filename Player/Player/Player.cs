using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Player.Interfaces;
using Player.Common;
using System.Linq;
using Player.Messages.Responses;
using Player.Messages.Requests;
using Player.GameObjects;
using System.Net.Sockets;
using System.Threading.Tasks;

namespace Player
{
    public class Player
    {
        public int Id;
        public int TeamId;
        public bool IsLeader;
        public string GameName;
        public int AskLevel;
        public int RespondLevel;
        public int Timeout;
        public string ServerHostName => _communicator.ServerHostName;
        public int ServerPort => _communicator.ServerPort;
        public IList<int> TeamMembersIds;
        public int LeaderId;
        public GameInfo Game;
        public List<Tile> Board = new List<Tile>();
        private ICommunicator _communicator;
        private IGameService _gameService;

        public Player(ICommunicator communicator, PlayerConfig config, IGameService gameService)
        {
            _communicator = communicator;
            _gameService = gameService;

            TeamId = config.TeamNumber;
            IsLeader = config.IsLeader;
            AskLevel = config.AskLevel;
            RespondLevel = config.RespondLevel;
            Timeout = config.Timeout;
            GameName = config.GameName;
        }

        public void Start()
        {
            GetGameInfo();
            ConnectToServer();
            WaitForGameStart();
            // RefreshBoardState(); -- gives us info about all teammates' (+ ours) initial position
            Play();
        }

        public void GetGameInfo()
        {
            var gamesList = _gameService.GetGamesList();
            Game = gamesList.FirstOrDefault(x => x.Name == GameName);
            if (Game == null)
            {
                throw new OperationCanceledException("Game not found");
            }

            for (int i = 0; i < Game.BoardSize.X * (Game.BoardSize.GoalArea * 2 + Game.BoardSize.TaskArea); i++)
            {
                Board.Add(new Tile());
            }
        }

        public void ConnectToServer()
        {
            if (!_communicator.IsConnected)
            {
                _communicator.Connect();
            }

            var helloMessage = new Message<PlayerHelloPayload>
            {
                Type = Consts.PlayerHelloRequest,
                SenderId = Consts.GameMasterId,
                Payload = new PlayerHelloPayload
                {
                    Game = GameName,
                    TeamId = TeamId,
                    IsLeader = IsLeader,
                    TemporaryId = new Random().Next(1, 10000)
                }
            };
            var helloMessageSerialized = JsonConvert.SerializeObject(helloMessage);
            _communicator.Send(helloMessageSerialized);


            Task<string> receivedMessageSerializedTask;
            try
            {
                receivedMessageSerializedTask = Task.Run(() => _communicator.Receive());
                if (!receivedMessageSerializedTask.Wait(Timeout))
                    throw new TimeoutException($"Did not receive any message after {Timeout}ms");
            }
            catch (AggregateException e)
            {
                throw e.InnerException;
            }
            var receivedMessageSerialized = receivedMessageSerializedTask.Result;

            var receivedGenericMessage = JsonConvert.DeserializeObject<Message>(receivedMessageSerialized);
            if (receivedGenericMessage.Type == Consts.PlayerRejected)
            {
                var rejectedMessage = JsonConvert.DeserializeObject<Message<PlayerRejectedPayload>>(receivedMessageSerialized);
                throw new PlayerRejectedException(rejectedMessage.Payload.Reason);
            }

            var acceptedMessage = JsonConvert.DeserializeObject<Message<PlayerAcceptedPayload>>(receivedMessageSerialized);
            Id = acceptedMessage.Payload.AssignedPlayerId;
        }

        public void Disconnect()
        {
            _communicator.Disconnect();
            Console.WriteLine("Player disconnected.");
        }
        public void WaitForGameStart()
        {
            string receivedMessageSerialized;
            while (true)
            {
                receivedMessageSerialized = _communicator.Receive();

                var receivedGenericMessage = JsonConvert.DeserializeObject<Message>(receivedMessageSerialized);

                if (receivedGenericMessage.Type == Consts.GameStarted) break;
            }

            var message = JsonConvert.DeserializeObject<Message<GameStartedPayload>>(receivedMessageSerialized);

            TeamMembersIds = message.Payload.TeamInfo[TeamId].Players;
            LeaderId = message.Payload.TeamInfo[TeamId].LeaderId;
        }

        public void Play()
        {
            while (true)
            {
                Discover();
                System.Threading.Thread.Sleep(10000);
            }
        }

        public bool Discover()
        {
            var message = new Message<DiscoveryPayload>()
            {
                Type = Consts.DiscoveryRequest,
                RecipientId = Consts.GameMasterId,
                SenderId = Id
            };
            var messageSerialized = JsonConvert.SerializeObject(message);
            _communicator.Send(messageSerialized);

            (bool actionResult, object data) = WaitForActionResult();
            if (!actionResult)
            {
                Console.WriteLine(data as String);
                return false;
            }

            var receivedSerialized = _communicator.Receive();
            var receivedRaw = JsonConvert.DeserializeObject<Message>(receivedSerialized);
            if (receivedRaw.Type != Consts.DiscoveryResponse)
                throw new InvalidTypeReceivedException($"Excpected: {Consts.DiscoveryResponse} Received: {receivedRaw.Type}");

            var received = JsonConvert.DeserializeObject<Message<DiscoveryResponsePayload>>(receivedSerialized);
            foreach (var tileDTO in received.Payload.Tiles)
            {
                var tile = Board[tileDTO.X + Game.BoardSize.X * tileDTO.Y];
                tile = AutoMapper.Mapper.Map<Tile>(tileDTO);
                tile.Timestamp = received.Payload.Timestamp;
                if (tileDTO.Piece)
                {
                    tile.Piece = new Piece();
                }
            }
            return true;
        }


        public (bool, object) WaitForActionResult()
        {
            var receivedSerialized = _communicator.Receive();
            var receivedRaw = JsonConvert.DeserializeObject<Message>(receivedSerialized);

            if (receivedRaw.Type == Consts.ActionValid)
            {
                var received = JsonConvert.DeserializeObject<Message<ActionValidPayload>>(receivedSerialized);
                return (true, received.Payload.Delay);

            }
            if (receivedRaw.Type == Consts.ActionInvalid)
            {
                var received = JsonConvert.DeserializeObject<Message<ActionInvalidPayload>>(receivedSerialized);
                return (false, received.Payload.Reason);
            }

            throw new InvalidTypeReceivedException($"Excpected: ACTION_VALID/INVALID Received: {receivedRaw.Type}");
        }
    }
}
