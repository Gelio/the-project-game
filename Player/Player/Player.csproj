<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <None Remove="player.config.json" />
  </ItemGroup>

  <ItemGroup>
    <Content Include="player.config.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="automapper" Version="6.2.2" />
    <PackageReference Include="Newtonsoft.Json" Version="11.0.1" />
    <PackageReference Include="NLog" Version="4.5.2" />
  </ItemGroup>

</Project>
