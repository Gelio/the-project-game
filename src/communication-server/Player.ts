import { LoggerInstance } from 'winston';

import { Communicator } from '../common/Communicator';

import { Message } from '../interfaces/Message';
import { Service } from '../interfaces/Service';

import { MessageRouter } from './MessageRouter';
import { PlayerInfo } from './PlayerInfo';

export class Player implements Service {
  public readonly info: PlayerInfo;

  private readonly communicator: Communicator;
  private readonly messageRouter: MessageRouter;
  private readonly logger: LoggerInstance;

  public get id() {
    return this.info.id;
  }

  constructor(
    communicator: Communicator,
    messageRouter: MessageRouter,
    logger: LoggerInstance,
    playerInfo: PlayerInfo
  ) {
    this.communicator = communicator;
    this.messageRouter = messageRouter;
    this.logger = logger;
    this.info = playerInfo;

    this.handleMessage = this.handleMessage.bind(this);
  }

  public init() {
    this.messageRouter.registerPlayerCommunicator(this.id, this.communicator);
    this.communicator.on('message', this.handleMessage);
  }

  public destroy() {
    this.logger.verbose(`Destroying player ${this.id}`);

    if (this.messageRouter.hasRegisteredPlayerCommunicator(this.id)) {
      this.messageRouter.unregisterPlayerCommunicator(this.id);
    }
  }

  private handleMessage<T>(message: Message<T>) {
    if (!this.isMessageValid(message)) {
      this.logger.warn(
        `Received message with sender ID ${message.senderId} but player ID is ${this.id}`
      );

      return;
    }

    this.messageRouter.sendMessageToGameMaster(this.info.gameName, message);
  }

  private isMessageValid<T>(message: Message<T>) {
    if (!this.id) {
      return true;
    }

    return this.id === message.senderId;
  }
}
