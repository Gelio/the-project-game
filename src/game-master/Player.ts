import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
