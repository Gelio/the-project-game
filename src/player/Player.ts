import { createConnection } from 'net';
import { LoggerInstance } from 'winston';

import { config } from './config';

import { bindObjectProperties } from '../common/bindObjectProperties';
import { Communicator } from '../common/Communicator';
import { LoggerFactory } from '../common/logging/LoggerFactory';
import { UITransport } from '../common/logging/UITransport';
import { TeamId } from '../common/TeamId';

import { Message } from '../interfaces/Message';
import { PlayerAcceptedMessage } from '../interfaces/messages/PlayerAcceptedMessage';
import { PlayerHelloMessage } from '../interfaces/messages/PlayerHelloMessage';
import { PlayerRejectedMessage } from '../interfaces/messages/PlayerRejectedMessage';
import { MessageWithRecipient } from '../interfaces/MessageWithRecipient';
import { ListGamesRequest } from '../interfaces/requests/ListGamesRequest';
import { ListGamesResponse } from '../interfaces/responses/ListGamesResponse';
import { Service } from '../interfaces/Service';

import { registerUncaughtExceptionHandler } from '../registerUncaughtExceptionHandler';

import { UIController } from './ui/UIController';

export interface PlayerOptions {
  serverHostname: string;
  serverPort: number;
  askLevel: number;
  respondLevel: number;
  teamNumber: TeamId;
  teamLeader: boolean;
  timeout: number;
}

export class Player implements Service {
  private id: number;
  // @ts-ignore
  private isAccepted = false;

  private options: PlayerOptions;

  private communicator: Communicator;
  private logger: LoggerInstance;
  private readonly uiController: UIController;
  private readonly loggerFactory: LoggerFactory;

  private messageHandlers: { [type: string]: Function } = {
    PLAYER_ACCEPTED: this.handlePlayerAccepted,
    PLAYER_REJECTED: this.handlePlayerRejected
  };

  constructor(options: PlayerOptions, uiController: UIController, loggerFactory: LoggerFactory) {
    this.options = options;

    this.uiController = uiController;
    this.loggerFactory = loggerFactory;

    bindObjectProperties(this.messageHandlers, this);
    this.handleMessage = this.handleMessage.bind(this);
    this.handleServerDisconnection = this.handleServerDisconnection.bind(this);
  }

  public init() {
    const { serverHostname, serverPort } = this.options;

    this.initUI();
    this.initLogger();

    const socket = createConnection(
      {
        host: serverHostname,
        port: serverPort
      },
      () => {
        this.logger.info(`Connected to the server at ${serverHostname}:${serverPort}`);
        this.sendListGames();
        this.sendHandshake();
      }
    );

    this.communicator = new Communicator(socket, this.logger);
    this.communicator.bindListeners();

    this.communicator.on('message', this.handleMessage);
    this.communicator.on('close', this.handleServerDisconnection);
  }

  public destroy() {
    this.communicator.destroy();
    this.uiController.destroy();
  }

  private handleServerDisconnection() {
    this.destroy();
    this.logger.info('Disconnected from the server');
  }

  private sendHandshake() {
    const temporaryId = Math.floor(Math.random() * config.maxTemporaryPlayerId);

    const message: PlayerHelloMessage = {
      type: 'PLAYER_HELLO',
      senderId: -2,
      payload: {
        isLeader: this.options.teamLeader,
        teamId: this.options.teamNumber,
        temporaryId,
        game: 'the project game'
      }
    };

    this.communicator.sendMessage(message);
  }

  private async sendListGames() {
    const message: ListGamesRequest = {
      type: 'LIST_GAMES_REQUEST',
      senderId: -2,
      payload: undefined
    };

    this.communicator.sendMessage(message);
    const listGamesResponse = <ListGamesResponse>await this.communicator.waitForSpecificMessage(
      (msg: MessageWithRecipient<ListGamesResponse>) => msg.type === 'LIST_GAMES_RESPONSE'
    );

    this.handleListGamesResponse(listGamesResponse);
  }

  private handleMessage<T>(message: Message<T>) {
    // @ts-ignore
    if (this.messageHandlers[message.type] !== undefined) {
      return this.messageHandlers[message.type](message);
    }
  }

  private handleListGamesResponse(message: ListGamesResponse) {
    const games = message.payload.games;
    for (const game of games) {
      this.logger.info(`\n Game \"${game.name}\" : \"${game.description}\"`);
      this.logger.info(
        `\n boardSize: x:${game.boardSize.x} taskArea:${game.boardSize.taskArea} goalArea: ${
          game.boardSize.goalArea
        }`
      );
      this.logger.info(
        `\n Goal limit:${game.goalLimit} teamSizes: 1:${game.teamSizes[1]} 2:${game.teamSizes[2]}`
      );
      this.logger.info(`\n ${JSON.stringify(game.delays)}`);
    }
  }

  private handlePlayerAccepted(message: PlayerAcceptedMessage) {
    this.id = message.payload.assignedPlayerId;
    this.isAccepted = true;

    this.logger.info(`Player accepted with id ${this.id}`);
  }

  private handlePlayerRejected(message: PlayerRejectedMessage) {
    this.logger.error(`Player rejected. Reason: ${message.payload.reason}`);
    this.destroy();
  }

  private initUI() {
    this.uiController.init();
    this.uiController.render();
  }

  private initLogger() {
    const uiTransport = new UITransport(this.uiController.log.bind(this.uiController));
    this.logger = this.loggerFactory.createLogger([uiTransport]);
    registerUncaughtExceptionHandler(this.logger);

    this.logger.info('Logger initiated');
  }
}
