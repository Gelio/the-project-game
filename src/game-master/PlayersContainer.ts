import { TeamId } from '../common/TeamId';
import { Player } from './Player';

export class PlayersContainer {
  public readonly players: Player[] = [];

  public addPlayer(player: Player) {
    if (this.players.indexOf(player) !== -1) {
      throw new Error('Player already added');
    }
    this.players.push(player);
  }

  public removePlayer(player: Player) {
    const playerIndex = this.players.indexOf(player);
    if (playerIndex === -1) {
      throw new Error('Player is not added');
    }
    this.players.splice(playerIndex, 1);
  }

  public getPlayersFromTeam(teamId: TeamId) {
    return this.players.filter(player => player.teamId === teamId);
  }

  public getConnectedPlayers() {
    return this.players.filter(player => player.isConnected);
  }

  public getPlayerById(playerId: number) {
    return this.players.find(player => player.playerId === playerId);
  }
}
