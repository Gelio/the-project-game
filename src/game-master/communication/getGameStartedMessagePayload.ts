import { PlayersContainer } from '../PlayersContainer';

import { GameStartedMessagePayload } from '../../interfaces/messages/GameStartedMessage';

export function getGameStartedMessagePayload(
  playersContainer: PlayersContainer
): GameStartedMessagePayload {
  const team1Players = playersContainer.getPlayersFromTeam(1);
  const team2Players = playersContainer.getPlayersFromTeam(2);
  const team1Leader = team1Players.find(player => player.isLeader);
  const team2Leader = team2Players.find(player => player.isLeader);

  if (!team1Leader || !team2Leader) {
    throw new Error('Game cannot start without both leaders');
  }

  return {
    teamInfo: {
      1: {
        players: team1Players.map(player => player.playerId),
        leaderId: team1Leader.playerId
      },
      2: {
        players: team2Players.map(player => player.playerId),
        leaderId: team2Leader.playerId
      }
    }
  };
}
