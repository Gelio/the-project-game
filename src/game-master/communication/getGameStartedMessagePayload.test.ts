import { Player } from '../Player';
import { PlayersContainer } from '../PlayersContainer';

import { getGameStartedMessagePayload } from './getGameStartedMessagePayload';

describe('getGameStartedMessagePayload', () => {
  let playersContainer: PlayersContainer;
  let player1: Player;
  let player2: Player;

  beforeEach(() => {
    playersContainer = new PlayersContainer();

    player1 = new Player();
    player1.playerId = 'p1';
    player1.teamId = 1;
    player1.isLeader = true;

    player2 = new Player();
    player2.playerId = 'p2';
    player2.teamId = 2;
    player2.isLeader = true;
  });

  it('should should throw an error when there is no leader from team 1', () => {
    playersContainer.addPlayer(player2);

    expect(() => getGameStartedMessagePayload(playersContainer)).toThrowErrorMatchingSnapshot();
  });

  it('should should throw an error when there is no leader from team 2', () => {
    playersContainer.addPlayer(player1);

    expect(() => getGameStartedMessagePayload(playersContainer)).toThrowErrorMatchingSnapshot();
  });

  it('should generate a message matching snapshot', () => {
    playersContainer.addPlayer(player1);
    playersContainer.addPlayer(player2);

    const payload = getGameStartedMessagePayload(playersContainer);

    expect(payload).toMatchSnapshot();
  });
});
