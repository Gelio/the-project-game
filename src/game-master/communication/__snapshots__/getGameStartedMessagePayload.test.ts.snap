// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGameStartedMessagePayload should generate a message matching snapshot 1`] = `
Object {
  "teamInfo": Object {
    "1": Object {
      "leaderId": "p1",
      "players": Array [
        "p1",
      ],
    },
    "2": Object {
      "leaderId": "p2",
      "players": Array [
        "p2",
      ],
    },
  },
}
`;

exports[`getGameStartedMessagePayload should should throw an error when there is no leader from team 1 1`] = `"Game cannot start without both leaders"`;

exports[`getGameStartedMessagePayload should should throw an error when there is no leader from team 2 1`] = `"Game cannot start without both leaders"`;
